CRM/Core/Payment/PaymentExtended.php

```diff
@@ -201,7 +201,7 @@ protected function getGoBackUrl($qfKey) {
   protected function getNotifyUrl($allowLocalHost = FALSE) {
     $url = CRM_Utils_System::url(
       'civicrm/payment/ipn/' . $this->formatted_transaction_id . '/' . $this->_paymentProcessor['id'],
-      NULL,
+      array('qfKey' => CRM_Utils_Request::retrieve('qfKey', 'String')),
```

Comment:
@eileenmcnaughton So the session key is required for sagepay because the IPN handler returns the redirect URL to sagepay when requested, rather than using the URL via the initial transfer offsite.  Any thoughts on how this could be handled?

Reply:
@mattwire I'm trying to remove those functions that have been migrated to core from this extension - that goBackUrl is one of them

Reply:
@eileenmcnaughton We're talking about `getNotifyUrl` and not `getGoBackUrl`.  However, I see the same applies.  Basically the session key (qfKey) is essential for sagepay because (uniquely) Sagepay does the final user redirect back from their site ONLY after the IPN has completed.  It needs the session key to be returned so that it can redirect the user to the CiviCRM payment confirmation page.  Any idea how we could gracefully override this in core?

Reply:
@mattwire normally the QFKey is stored in the user session before leaving the site & retrieved from there on return?

Reply:
Right, so discussion on mattermost - we should see if we can modify "next url" so that it is set to `civicrm/payment/ipn/5` or equivalent and then we would not need the session key.

Reply:
Yep - although looking at the above - there is an Omnipay-specific thing already in that function 
- 'civicrm/payment/ipn/' . $this->formatted_transaction_id . '/'

Hmm - we shouldn't be sending next url to Sagepay but returnUrl - that will figure out nextUrl

Reply:
@eileenmcnaughton I don't think there is any way around putting the qfKey in getNotifyUrl:

Done a bit of debugging as I couldn't remember exactly what was going on.

**No session key supplied via getNotifyUrl (standard):**
1. CiviCRM payment submitted.
2. Goes offsite to Sagepay.
3. Sagepay returns via ipn (eg. civicrm/payment/ipn/146/5).
4. CiviCRM server responds to IPN via `processPaymentNotification` and then `redirectOrExit`.  Both of these functions are OUTSIDE of the session so `getStoredUrl` in `redirectOrExit` fails.
5. Sagepay receives redirect URL via `$response->confirm($redirectUrl, $userMsg)` and redirects the users browser to that URL, but the URL is invalid because we didn't know the session key.

**Session key supplied via getNotifyUrl (modified):**
1. CiviCRM payment submitted.
2. Goes offsite to Sagepay.
3. Sagepay returns via ipn (eg. civicrm/payment/ipn/146/5?qfKey=abc1234).
4. CiviCRM server responds to IPN via `processPaymentNotification` and then `redirectOrExit`.  Both of these functions are OUTSIDE of the session so `getStoredUrl` in `redirectOrExit` fails.  But we then try again, retrieving the session key from the request URL and now we have access to the session so it works.
5. Sagepay receives redirect URL via `$response->confirm($redirectUrl, $userMsg)` and redirects the users browser to that URL which is the correct URL because it came from the stored session.